p11-kit/generate-keypair.c

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (c) 2023, Red Hat Inc.
+ *
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above
+ *       copyright notice, this list of conditions and the
+ *       following disclaimer.
+ *     * Redistributions in binary form must reproduce the
+ *       above copyright notice, this list of conditions and
+ *       the following disclaimer in the documentation and/or
+ *       other materials provided with the distribution.
+ *     * The names of contributors to this software may not be
+ *       used to endorse or promote products derived from this
+ *       software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ *
+ * Author: Zoltan Fridrich <zfridric@redhat.com>
+ */
+
+#include "config.h"
+
+#include "attrs.h"
+#include "buffer.h"
+#include "compat.h"
+#include "debug.h"
+#include "iter.h"
+#include "message.h"
+#include "mock.h"
+#include "tool.h"
+
+#include <assert.h>
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef ENABLE_NLS
+#include <libintl.h>
+#define _(x) dgettext(PACKAGE_NAME, x)
+#else
+#define _(x) (x)
+#endif
+
+#define cleanup_if_fail(X, S)          \
+	do { if (X) { } else {         \
+		p11_message (_(S));    \
+		goto cleanup;          \
+	} } while (false)
+
+#define cleanup_if_fail_rv(X, S, R)    \
+	do { if (X) { } else {         \
+		p11_message (_(S), R); \
+		goto cleanup;          \
+	} } while (false)
+
+const CK_BBOOL TVAL = CK_TRUE;
```

Comment:
As `TVAL` and `FVAL` are only used in `get_templates`, can we make these as a local variable? Otherwise `static` is missing.

Reply:
If I make them into local variables inside `get_templates` then after return the addresses become invalid and I don't want to add them as additional arguments to `get_templates`. I will make them `static`.

Reply:
In that case I would use `p11_attrs` API, which internally keeps a copy. That would also make `get_templates` simpler at a cost of some heap allocations (it's not a problem here as this is not a performance critical code).

Reply:
I have added a commit with second `get_templates` function where I used the `attrs.h` API. I dislike it.

Reply:
I don't understand why you dislike it; the amount of code is smaller and you don't need to maintain attribute index by yourself (i and j).